soccer/vision/filter/KalmanFilter2D.cpp

```diff
@@ -0,0 +1,116 @@
+#include "KalmanFilter2D.hpp"
+
+#include <cmath>
+
+#include "vision/util/VisionFilterConfig.hpp"
+
+REGISTER_CONFIGURABLE(KalmanFilter2D)
+
+ConfigDouble* KalmanFilter2D::ball_init_covariance;
+ConfigDouble* KalmanFilter2D::ball_process_noise;
+ConfigDouble* KalmanFilter2D::ball_observation_noise;
+
+void KalmanFilter2D::createConfiguration(Configuration* cfg) {
+    ball_init_covariance = new ConfigDouble(cfg, "VisionFilter/Ball/init_covariance", 100);
+    ball_process_noise = new ConfigDouble(cfg, "VisionFilter/Ball/process_noise", .1);
+    ball_observation_noise = new ConfigDouble(cfg, "VisionFilter/Ball/observation_noise", 2.0);
+}
+
+KalmanFilter2D::KalmanFilter2D() : KalmanFilter(1,1) {}
+
+KalmanFilter2D::KalmanFilter2D(Geometry2d::Point initPos, Geometry2d::Point initVel)
+    : KalmanFilter(4, 2) {
+
+    // States are X pos, X vel, Y pos, Y vel
+    x_k1_k1 << initPos.x(),
+               initVel.x(),
+               initPos.y(),
+               initVel.y();
+    x_k_k1 = x_k1_k1;
+    x_k_k = x_k1_k1;
+
+    // Initial covariance is usually extremely high to converge to the true solution
+    double p = *ball_init_covariance;
+    P_k1_k1 << p, 0, 0, 0,
+               0, p, 0, 0,
+               0, 0, p, 0,
+               0, 0, 0, p;
+    P_k_k1 = P_k1_k1;
+    P_k_k = P_k1_k1;
+
+    // State transition matrix (A)
+    // Pos, velocity integrator. Assume constant velocity
+    double dt = *VisionFilterConfig::vision_loop_dt; // TODO: Take config values
+    F_k << 1, dt,  0,  0,
+           0,  1,  0,  0,
+           0,  0,  1, dt,
+           0,  0,  0,  1;
+
+    // Control transition matrix (B)
+    // No inputs
+    B_k << 0,
+           0,
+           0,
+           0;
+
+    // Observation Matrix (C)
+    // We can get positions
+    H_k << 1, 0, 0, 0,
+           0, 0, 1, 0;
+
+    // Covariance of process noise (how wrong A is)
+    // Based on a guassian white noise w_k in x_dot = A*x + B*u + G*w
+    // The noise can be propogated through the model resulting in a process noise
+    // of the form
+    //
+    //  [1/3 T^3     1/2 T^2] * sigma^2
+    //  [1/2 T^2           T]
+    // Where sigma is the standard deviation of the process noise
+    // the change in velocity over one time step should be around sqrt(T * sigma^2)
+    // Note: T is the sample period
+    // Taken from Tiger's AutoRef. Most likely found through integration of error through the
+    // state matrices
```

Comment:
More information here: https://en.wikipedia.org/wiki/Discretization#Discretization_of_process_noise. This is basically just an expansion of the matrix exponential.